x-pack/filebeat/input/httpjson/json_parser.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package httpjson
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// parse returns array of string values from JSON string
+//
+// Examples:
+// input:
+//   rawJSON={"a":[{"b":"b_value_1"},{"b":"b_value_2"},{"b":"b_value_3"}]}
+//   key=a.#.b
+// output:
+//   ["a_value_1", "a_value_2"]
+// input:
+//   rawJSON=[{"a":"a_value_1"},{"a":"a_value_2"}]
+//   key=#.a
+// output:
+//   ["a_value_1", "a_value_2"]
+func parse(rawJSON, key string) ([]string, error) {
+	var data interface{}
+	err := json.Unmarshal([]byte(rawJSON), &data)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshal input data: %w", err)
+	}
+	values, err := parseInterface(data, key)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing JSON: %w", err)
+	}
+	return values, nil
+}
+
+// parseInterface returns array of string values from JSON data interface
+//
+// Examples:
+// input:
+// 	 data=
+//   {
+//   	"a": {
+//   		"b": "b_value",
+//   	},
+//   }
+//   key=a..b
+// output:
+//   ["b_value"]
+// input:
+//   data=
+// 	 {
+// 	 	"a": [
+// 	 		{
+// 	 			"b":"b_value_1"
+// 	 		},
+// 	 		{
+// 	 			"b":"b_value_2"
+// 	 		}
+// 	 	]
+// 	 }
+//   key=a.#.b
+// output:
+//   ["b_value_1", "b_value_2"]
+func parseInterface(data interface{}, key string) (values []string, err error) {
+	splitKey := strings.Split(key, ".")
+	for i, key := range splitKey {
+		switch key {
+		case "#":
+			// '#' means it's an array and we want to get the
+			// next key from each object in the array
```

Comment:
This appears to be new syntax for array indexing. Would it be worth considering something like JSONPath (or a subset of that) that already exists and has more power?

Reply:
well, I will look into it.
I am not sure exactly but I think we can use jsonPATH.

Reply:
Hey, I looked into it and found some existing libraries available for [JSONPath](https://pkg.go.dev/github.com/yalp/jsonpath).

As we discussed internally to not use any of such external libraries to process JSON, rather than to construct our own custom logic, and that's why we haven't used any of such libraries like [gjson](https://github.com/tidwall/gjson) or [JSONPath](https://pkg.go.dev/github.com/yalp/jsonpath).

@akshay-saraswat @mtojek

Reply:
If this is just about a single function and we won't use 90% of any JSONPath library, then I would stick with a tiny custom implementation.

On the other hand, as we have @belimawr here, who is familiar with filebeat inputs, I would pass this question.

Reply:
My main concern was inventing a new syntax.

Reply:
https://github.com/elastic/beats/issues/22590 makes me believe this chain requests will be used in other places, so I see the possibility of this being used in other places.

If implementing (or adding a dependency) the full JSONPath is too much at the moment, we could use a subset of it, so later on if we need to extend this functionality we can just support the full JSONPath specification (thanks @cmacknz for this idea).

Anyway, I'm still reaching out to people so I can get a better high level view of this chain requests feature.

Reply:
Putting this here in case it is helpful. There is an expression-extensible JSONPath [implementation for Go](https://github.com/PaesslerAG/jsonpath). This takes expressions using an [extensible language system](https://github.com/PaesslerAG/gval). [Example code](https://pkg.go.dev/github.com/PaesslerAG/jsonpath?utm_source=godoc#example-package--Gval).